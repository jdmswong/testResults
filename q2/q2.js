/*
Return a new object with the following keys calculated and filled (about
  the persons data set)
totalRecordCount - integer - the total count of all persons records
datesOfBirth - object - another object, with year number (1986) as the key
  and the count of how many such records share the same year number as the
  value
genders - object - another object, similar to the one above, but each 
  of these will have the corresponding sum of genders (male / female)
smokers - object - another object, similar to the one above, but each
  of these will have the corresponding sum of smokers / non-smokers
 */

const reduceOnProperty = (array, property, transformer = x => x) =>
  array.reduce(
    (result, obj) => {
      const propValue = transformer(obj[property]);
      if (result[propValue]) {
        result[propValue] += 1;
      } else {
        result[propValue] = 1;
      }
      return result;
    }, {}
  );

export default persons => ({
  totalRecordCount: persons.length,
  datesOfBirth: reduceOnProperty(
    persons, 'DateOfBirth', birthDate => birthDate.match(/\d{4}/)[0]
  ),
  genders: reduceOnProperty(persons, 'Gender'),
  smokers: reduceOnProperty(persons, 'Smoker'),
});
