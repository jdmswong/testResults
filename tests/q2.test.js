import q2 from '../q2';
import perons from '../persons';

test('empty set', () => {
  const sampleData = [];
  expect(q2(sampleData)).toMatchSnapshot();
});

test('returns expected output', () => {
  expect(q2(perons)).toMatchSnapshot();
});

