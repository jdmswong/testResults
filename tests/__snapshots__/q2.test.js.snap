// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty set 1`] = `
Object {
  "datesOfBirth": Object {},
  "genders": Object {},
  "smokers": Object {},
  "totalRecordCount": 0,
}
`;

exports[`returns expected output 1`] = `
Object {
  "datesOfBirth": Object {
    "1948": 1,
    "1955": 1,
    "1964": 1,
    "1970": 1,
    "1974": 2,
    "1987": 1,
    "2000": 3,
    "2001": 2,
    "2004": 1,
    "2017": 1,
  },
  "genders": Object {
    "Female": 7,
    "Male": 7,
  },
  "smokers": Object {
    "false": 9,
    "true": 5,
  },
  "totalRecordCount": 14,
}
`;
